stripe/src/main/java/com/stripe/android/IssuingCardPinService.java

```diff
@@ -0,0 +1,274 @@
+            @NonNull EphemeralKeyProvider keyProvider
+    ) {
+        mEphemeralKeyManager = new EphemeralKeyManager<>(
+                keyProvider,
+                this,
+                KEY_REFRESH_BUFFER_IN_SECONDS,
+                null,
+                IssuingCardEphemeralKey.class);
+    }
+
+    /**
+     * Create a IssuingCardPinService with the provided {@link EphemeralKeyProvider}.
+     *
+     * @param keyProvider an {@link EphemeralKeyProvider} used to get
+     *                    {@link CustomerEphemeralKey EphemeralKeys} as needed
+     */
+    @NonNull
+    public static IssuingCardPinService create(@NonNull EphemeralKeyProvider keyProvider) {
+        return new IssuingCardPinService(keyProvider);
+    }
+
+    /**
+     * Retrieves a PIN for a given card
+     *
+     * @param cardId          the ID of the card (looks like ic_abcdef1234)
+     * @param verificationId  the ID of the verification that was sent to the cardholder
+     *                        (typically server-side, through /v1/issuing/verifications)
+     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
+     * @param listener        a listener for either the PIN, or any error that can occur
+     */
+    public void retrievePin(
+            @NonNull String cardId,
+            @NonNull String verificationId,
+            @NonNull String userOneTimeCode,
+            @Nullable IssuingCardPinRetrievalListener listener
+    ) {
+
+        Map<String, Object> arguments = new HashMap<>();
+        arguments.put("cardId", cardId);
+        arguments.put("verificationId", verificationId);
+        arguments.put("userOneTimeCode", userOneTimeCode);
+        arguments.put("listener", listener);
+
+        mEphemeralKeyManager.retrieveEphemeralKey(PIN_RETRIEVE, arguments);
+    }
+
+    /**
+     * Retrieves a PIN for a given card
+     *
+     * @param cardId          the ID of the card (looks like ic_abcdef1234)
+     * @param newPin          the new desired PIN
+     * @param verificationId  the ID of the verification that was sent to the cardholder
+     *                        (typically server-side, through /v1/issuing/verifications)
+     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
+     * @param listener        a listener for either the PIN, or any error that can occur
+     */
+    public void updatePin(
+            @NonNull String cardId,
+            @NonNull String newPin,
+            @NonNull String verificationId,
+            @NonNull String userOneTimeCode,
+            @Nullable IssuingCardPinUpdateListener listener
+    ) {
+
+        Map<String, Object> arguments = new HashMap<>();
+        arguments.put("cardId", cardId);
+        arguments.put("newPin", newPin);
+        arguments.put("verificationId", verificationId);
+        arguments.put("userOneTimeCode", userOneTimeCode);
+        arguments.put("listener", listener);
```

Comment:
I think a strong reference to the listener can potentially result in a memory leak (e.g. the `IssuingCardPinUpdateListener` can hold a reference to an activity, and if that activity is going away we'll still hold a reference to it here).

What do you think of changing this to
`arguments.put("listener", new WeakReference<>(listener));`

and downstream as well

Reply:
ah yes